### PR TITLE
Transition Laptop Request Status (2)

### PR DESCRIPTION
**Pull request by Fareedh-12**
## Naming Rules

Name your PR like this: ISSUENUMBER-TITLE-YOURNAME

## Description

Write a summary of your change. Explain why you have made it.

## Related to
laptop request status functionality

Make sure you include the issue number with a hash sign # so Github can link this PR to the right issue, like this:

Fixes #1
Updated endpoint: GET /api/laptop_request/:id
Updated endpoint: PUT /api/laptop_assignment/:id
Used endpoint: GET /api/laptop_donation/:id

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have carefully reviewed my own code
- [x] I have commented my code
- [x] I have updated any documentation

### CODE REVIEW

**Comment by Fareedh-12** [comment]
@gregdyke Please review my changes

File: client/src/pages/ViewRequestStatus/RequestStatusView.js [hunk 1]
```diff
@@ -1,20 +1,77 @@
 import React, { useEffect, useState } from "react";
+// import { Navigate } from "react-router-dom";
 
 import { useParams } from "react-router-dom";
 import ButtonComponent from "./ButtonComponent";
 
 export function RequestStatus() {
 	const [request, setRequest] = useState(null);
+	const [donation, setDonation] = useState(null);
+	const [isCancelled, setIsCancelled] = useState(false);
```
Comment (gregdyke) [inline client/src/pages/ViewRequestStatus/RequestStatusView.js]: probably you just needed a needsReloading state, along with a setNeedsReloading, but this is fine too

Comment (gregdyke) [reply client/src/pages/ViewRequestStatus/RequestStatusView.js]: (instead of different states for the isCancelled, isRejected, etc transition)

**Review by gregdyke — APPROVED** [review]
Nice job. Looks good to go like this (I assume you tested locally). Good job figuring out the bugs you must have had (ACCEPTED vs accepted, broken put request, etc.)

File: client/src/pages/ViewRequestStatus/RequestStatusView.js (hunk 1 above)
Comment (Fareedh-12) [reply client/src/pages/ViewRequestStatus/RequestStatusView.js]: Done that

File: client/src/pages/Home/Forms/DonatorFormReqest/DonatorForm.js [hunk 2]
```diff
@@ -102,20 +102,11 @@ function DonatorForm() {
 							<input
 								type="radio"
 								name="deliveryOption"
-								value="PICK UP"
+								value="PICKßUP"
```
Comment (gregdyke) [inline client/src/pages/Home/Forms/DonatorFormReqest/DonatorForm.js]: This looks wrong, PICKUP?

File: client/src/pages/ViewRequestStatus/RequestStatusView.js [hunk 3]
```diff
@@ -1,20 +1,77 @@
...
+			}),
+			headers: { "content-type": "application/json" },
+		}).then(() => {
+			setNeedsReloading(!needsReloading);
+		});
+	};
+
+	const rejectRequest = () => {
+		fetch(`/api/laptop_assignment/${request.assignmentId}`, {
+			method: "delete",
```
Comment (gregdyke) [inline client/src/pages/ViewRequestStatus/RequestStatusView.js]: DELETE

File: server/api.js [hunk 4]
```diff
@@ -211,7 +165,34 @@ router.post("/laptop_donation", (req, res) => {
 		email,
 		deliveryOption,
 	])
-		.then(() => res.send("result.rows"))
+		.then(async () => {
+			let laptopDonationResult = await db.query(
```
Comment (gregdyke) [inline server/api.js]: This is wrong. You already have a laptop_donation: the one you just inserted

File: server/api.js [hunk 5]
```diff
@@ -211,7 +165,34 @@ router.post("/laptop_donation", (req, res) => {
...
+			);
+			let laptopDonation = {};
+			if (laptopDonationResult.rows.length > 0) {
+				laptopDonation = {
+					id: laptopDonationResult.rows[0].id,
+				};
+			}
+			if (laptopDonation.id) {
+				const requestIDResult = await db.query(
+					"SELECT id FROM laptop_request where id is not null AND id NOT IN (SELECT laptop_request_id FROM laptop_assignment)"
```
Comment (gregdyke) [inline server/api.js]: what does "where id is not null" get you?

File: server/api.js [hunk 6]
```diff
@@ -211,7 +165,34 @@ router.post("/laptop_donation", (req, res) => {
...
+				);
+				let requestID = {};
```
Comment (gregdyke) [inline server/api.js]: requestId, not requestId, but why is it an object and not just an id?

File: server/api.js (hunk 4 above)
Comment (gregdyke) [inline server/api.js]: the mixture of then and await is a mess. Use one or the other (maybe in a followup pull request though)

File: server/api.js [hunk 7]
```diff
@@ -211,7 +165,34 @@ router.post("/laptop_donation", (req, res) => {
...
+				if (requestIDResult.rows.length > 0) {
+					requestID = {
+						id: requestIDResult.rows[0].id,
+					};
+				}
+				if (requestID.id) {
+					const assignmentQuery =
```
Comment (gregdyke) [inline server/api.js]: you have two places where you are creating an assignment. In a followup pull request, extract into a single function

**Review by gregdyke — CHANGES_REQUESTED** [review]

File: client/src/pages/Home/Forms/DonatorFormReqest/DonatorForm.js (hunk 2 above)
Comment (Fareedh-12) [reply client/src/pages/Home/Forms/DonatorFormReqest/DonatorForm.js]: Done that

File: client/src/pages/ViewRequestStatus/RequestStatusView.js (hunk 3 above)
Comment (Fareedh-12) [reply client/src/pages/ViewRequestStatus/RequestStatusView.js]: Done that

File: server/api.js (hunk 4 above)
Comment (Fareedh-12) [reply server/api.js]: @gregdyke 
Done that

File: server/api.js (hunk 5 above)
Comment (Fareedh-12) [reply server/api.js]: Removed not null

File: server/api.js (hunk 6 above)
Comment (Fareedh-12) [reply server/api.js]: using just the id

File: server/api.js (hunk 4 above)
Comment (Fareedh-12) [reply server/api.js]: I think it is necessary to use both, I could demo why I think that

File: server/api.js (hunk 7 above)
Comment (Fareedh-12) [reply server/api.js]: I don't think I quite follow what you mean by that.

**Review by gregdyke — APPROVED** [review]
Very nice work overall!

**Pull request merged**